### DISCUSSION TITLE
[Quicker 动作]批量下载 Markdown 中的外链图片并替换成本地离线图片链接（主要应用场景：通过简悦导出 Markdown 到 Obsidian 中）

### DISCUSSION THREAD

**Discussion by 1wingedangel** [Kenshin/simpread#2241] [服务]
(2021-10-14 更新) Obsidian 插件市场里有人放出了类似的插件：[aleksey-rezvov/obsidian-local-images](https://github.com/aleksey-rezvov/obsidian-local-images) 如果没什么特殊需求就请你们使用那个插件吧……
(2021-10-09 更新) 本脚本进行了大幅度的更新，原有的说明不再适用，请按照新版进行下载和安装，并且按照新版的说明进行操作。原来的版本将作为存档放在最后。

简悦提供了阅读模式以及标注的 Markdown 格式输出和离线 Markdown 格式输出。 Markdown 格式输出的时候，文章中的图片是以外链的形式存在，如：

```
![img](https://www.gamespark.jp/imgs/zoom/488339.png)
```

这种形式并不适合于离线的文章保存，所以如果要离线保存，我们可能需要导出离线 Markdown 格式。而离线 Markdown 格式则把图片用 Base64 的形式写在文档中，在 Obsidian 等 Markdown 编辑器上编辑的时候此类的内嵌图片形式会影响二次编辑和文档加载的时间。

所以我试着通过用 [Quicker](https://getquicker.net) 写个自动脚本来解决这个问题。
### 实现原理

1. 获取选中文本。没有选中文本的话从剪贴板读取内容。
2. 逐行判断有无外链图片的语法。
3. 如果找到了外链图片，提取图片外链地址，下载到本地。
4. 离线图片的位置为第一次运行脚本时指定的文件夹下 `assets` → `笔记名称` 的文件夹中。文件命名方式为 `yyyyMMddHHmmss_ffff.{图片文件扩展名}` 。`笔记名称` 通过模拟Obsidian的 `复制当前文件的路径` 的快捷键获取，若没有获取成功则弹出输入框让用户手动输入。
5. 把外链图片地址替换成本地离线图片文件的相对位置。
6. 合并生成新的文本。之前选中文本的话直接输出并覆盖原来选中的文本内容。如果没有选中文本的话输出到剪贴板。

### 前提

- 可不仅限于简悦导出的 Markdown 格式文档。
- 需要安装 [Quicker](https://getquicker.net) ，最好是更新为最新版本。
- Quicker 应该不需要专业版账户，免费版应该也能运行。（待验证）
- 需要预先在Obsidian里设置 `复制当前文件的路径` 快捷键。请设置为 `Ctrl+Shift+Y` 。（可以通过编辑脚本自己定制）
- 如果玩 Quicker 比较熟悉的用户，可根据自己需要适当修改脚本内容来满足自己的需要。

#### 设置 Obsidian 快捷键

- 进入 Obsidian 的 `设置` → `快捷键`
- 搜索 `复制当前文件的路径` 并设置为 `Ctrl+Shift+Y` 。

#### 安装 Quicker 脚本

访问地址：https://getquicker.net/sharedaction?code=16d0410b-2e88-4e30-91fb-08d982821308 ，并点击右侧的 `复制到剪贴板` 。

![Snipaste_2021-05-24_12-27-40](https://user-images.githubusercontent.com/396190/119298510-c10afc80-bc8f-11eb-8ccb-65bd33460a8d.png)

打开 Quicker 的面板，在任意空白按钮上右键，选择 `粘贴分享的动作` 。

![Snipaste_2021-05-24_12-31-36](https://user-images.githubusercontent.com/396190/119298532-c9633780-bc8f-11eb-9788-a982bb44241a.png)

选择 `安装` 即可。

### 使用脚本

这个脚本有两种使用方式。

#### 选中文本→处理替换→覆盖

1. 选中含有外链图片的 Markdown 文本片段。
2. 打开 quicker 面板，执行脚本。
3. 外链图片地址将会覆盖成本地离线图片相对位置。
4. 修改好的 Markdown 文本片段将会自动覆盖选中的文本。

#### 复制文本到剪贴板→处理替换→手动粘贴

1. 复制 Markdown 文本片段。
2. 打开 Quicker 面板，执行脚本。
3. 外链图片地址将会覆盖成本地离线图片相对位置。
4. 修改好的 Markdown 文本片段将会自动覆盖剪贴板内容。
5. 在任意位置粘贴剪贴板中的 Markdown 文本片段。

### 第一次运行脚本

- 第一次运行脚本时会提示指定 Obsidian 的文件夹。请指定 **Obsidian的Vault文件夹** 。
- 第一次运行脚本之后会在本地保存状态数据（具体位置：`C:\User\yourusername\AppData\Local\Quicker\states\state_c2d4018d-1257-4b58-b5db-d253c988cda4.json`），该状态在各个设备上独立，也就是说每台设备第一次运行都要设置各自的 Obsidian Vault 文件夹。

### 如果没有设置快捷键

- 如果没有设置 Obsidian 的快捷键，则会提示用户手动输入 `笔记名称` 。因此本脚本其实也可以用于非 Obsidian 的编辑器，但是不保证脚本的正常运行。（主要是工作目录相关）

#### 使用效果

使用脚本之前：

![Snipaste_2021-10-09_13-47-39](https://user-images.githubusercontent.com/396190/136645880-b9627931-dd82-451b-9211-cd7fb9cb5e6c.png)

使用脚本之后：

![Snipaste_2021-10-09_13-48-17](https://user-images.githubusercontent.com/396190/136645885-25fe80c4-e948-4ea5-9895-2b518ce4d6b8.png)

图片将会下载到该笔记目录的 `Assets\笔记名称` 文件夹下：

![Snipaste_2021-10-09_13-49-13](https://user-images.githubusercontent.com/396190/136645903-b07bb739-4f4b-47c3-8303-b8b8af10b389.png)

### 使用上的注意

- 由于本地离线图床位置固定，所以当你移动 Markdown 文档至新的位置时，图片显示可能会失效。在 Obsidian 的 Vault中移动 Markdown 文档不会有影响，但是为了保证兼容其他 Markdown 的编辑器，尽量少移动 Markdown 文档。
- 为了方便迁移文档数据或是打包分享的情况，特地把图片文件夹设置成 `Assets\笔记名称` 的形式。
- 目前我只支持了 `jpg|jpeg|png|gif` 格式的图片。如果有其他格式需求请自行修改脚本添加。
- 有什么使用上的建议请积极反馈。

----
### 实现原理

1. 获取选中文本。没有选中文本的话从剪贴板读取内容。
2. 逐行判断有无外链图片的语法。
3. 如果找到了外链图片，提取图片外链地址，下载到本地。
4. 离线图片的位置为**环境变量所指定的位置**下的 `assets` 文件夹中。文件命名方式为 `yyyyMMddHHmmss_ffff.{图片文件扩展名}` 。
5. 把外链图片地址替换成本地离线图片文件的相对位置。
6. 合并生成新的文本。之前选中文本的话直接输出并覆盖原来选中的文本内容。如果没有选中文本的话输出到剪贴板。

### 前提

- 可不仅限于简悦导出的 Markdown 格式文档。
- 需要安装 [Quicker](https://getquicker.net) ，最好是更新为最新版本。
- Quicker 应该不需要专业版账户，免费版应该也能运行。（待验证）
- 为支持多设备不同本地目录的需求，需要设置系统环境变量来指定下载目录。
- 如果玩 Quicker 比较熟悉的用户，可根据自己需要适当修改脚本内容来满足自己的需要。

#### 设置系统环境变量

- Win 键启动开始菜单 → `查看高级系统设置` → `环境变量`
- 创建新的用户变量或系统变量，变量名为 `OBWEB` ，变量值为**存放 Markdown 文档的路径** ，如下图。

![Snipaste_2021-05-24_09-50-46](https://user-images.githubusercontent.com/396190/119298499-bc464880-bc8f-11eb-9cff-c9b97d70c380.png)

> 增加环境变量之后可能需要重启 Quicker 才生效。

#### 安装 Quicker 脚本

访问地址：https://getquicker.net/sharedaction?code=c1799420-32c0-40a4-08aa-08d91c1f78c6 ，并点击右侧的 `复制到剪贴板` 。

![Snipaste_2021-05-24_12-27-40](https://user-images.githubusercontent.com/396190/119298510-c10afc80-bc8f-11eb-8ccb-65bd33460a8d.png)

打开 Quicker 的面板，在任意空白按钮上右键，选择 `粘贴分享的动作` 。

![Snipaste_2021-05-24_12-31-36](https://user-images.githubusercontent.com/396190/119298532-c9633780-bc8f-11eb-9788-a982bb44241a.png)

选择 `安装` 即可。

### 使用脚本

这个脚本有两种使用方式。

#### 选中文本→处理替换→覆盖

1. 选中含有外链图片的 Markdown 文本片段。
2. 打开 quicker 面板，执行脚本。
3. 外链图片地址将会覆盖成本地离线图片相对位置。
4. 修改好的 Markdown 文本片段将会自动覆盖选中的文本。

#### 复制文本到剪贴板→处理替换→手动粘贴

1. 复制 Markdown 文本片段。
2. 打开 Quicker 面板，执行脚本。
3. 外链图片地址将会覆盖成本地离线图片相对位置。
4. 修改好的 Markdown 文本片段将会自动覆盖剪贴板内容。
5. 在任意位置粘贴剪贴板中的 Markdown 文本片段。

#### 使用效果

使用脚本之前：

![Snipaste_2021-05-24_12-40-44](https://user-images.githubusercontent.com/396190/119298545-d08a4580-bc8f-11eb-9e38-de3c25169c70.png)

使用脚本之后：

![Snipaste_2021-05-24_12-43-29](https://user-images.githubusercontent.com/396190/119298557-d5e79000-bc8f-11eb-9acc-e8ee03c2957c.png)

### 使用上的注意

- 由于本地离线图床位置固定，所以当你移动 Markdown 文档至新的位置时，图片显示可能会失效。在 Obsidian 的 Vault中移动 Markdown 文档不会有影响，但是为了保证兼容其他 Markdown 的编辑器，尽量少移动 Markdown 文档。
- 增加环境变量之后运行脚本，提示没有设置环境变量的，请重启一下 Quicker 之后再试。
- 目前我只支持了 `jpg|jpeg|png|gif` 格式的图片。如果有其他格式需求请自行修改脚本添加。
- 有什么使用上的建议请积极反馈。

**Comment by YB-Kang** [comment]
请教下，转换成本地图片后，用obsidian怎么直接预览图片，因为Obsidian支持预览本地图片的格式是`![[]]`

**Reply by 1wingedangel** [reply]
> 请教下，转换成本地图片后，用obsidian怎么直接预览图片，因为Obsidian支持预览本地图片的格式是`![[]]`

Obsidian可以通过 `![](imageURL)` 来显示预览本地图片。其中 `imageURL` 可以是：

* 图片文件的绝对路径。
* 图片文件名。（前提：在Vault里此文件名是唯一的）
* 部分图片文件路径。比如 `1.png` 在Vault中任意一个 `Assets` 文件夹中，则 `Assets/1.png` 就能识别得到。

`![[]]` 这个情况是你通过附件的形式添加的图片才会是这个格式，这个不在此脚本的讨论范围内。

**Reply by Fung-NG** [reply]
obsidian是![ ]   
typora是![img]    能请教叫下这个quicker里应该如何修改吗
!\[[^\]]*\]\((http(s)?:\/\/.+?\.(jpg|jpeg|png|gif))\)
我改成这个不行

**Reply by 1wingedangel** [reply]
我最近在更新这个脚本以符合我的最新需求（现在我开始结合obsidian + logseq + orgzly进行记录），有时间我会把这篇文章和脚本内容改一下，到时候应该可以兼容typora。

另外你的问题可能是在正则表达式上。建议你用一些正则表达式测试网站测试一下。

**Reply by Fung-NG** [reply]
我也是用logseq typora只是一个中转站 网页的内容直接复制到logseq不太好 中转后会好很多，特别是幕布这种格式  楼主可以试试，可以保留很多东西，不知楼主是如何剪藏到logseq的

**Reply by Kenshin** [reply]
> 我也是用logseq typora只是一个中转站 网页的内容直接复制到logseq不太好 中转后会好很多，特别是幕布这种格式 楼主可以试试，可以保留很多东西，不知楼主是如何剪藏到logseq的

你可以试试

- https://github.com/Kenshin/simpread/discussions/2852

**Reply by 1wingedangel** [reply]
> 我也是用logseq typora只是一个中转站 网页的内容直接复制到logseq不太好 中转后会好很多，特别是幕布这种格式 楼主可以试试，可以保留很多东西，不知楼主是如何剪藏到logseq的

为了和Android上的 Orgzly （基于Org mode 的笔记与任务管理应用）紧密结合，我 Logseq 用的是 Org 模式。我发现就算是在 Org 模式下，logseq 的 graph 中的md文件还是能够正常识别只是无法编辑（编辑之后会变成 org/md 语法混合的奇葩文档），所以我试着把 logseq 的 Graph 放在 Obsidian 的 Vault 里，然后 Obsidian 的默认笔记路径改成 logseq 的 Pages 文件夹。这样每次我在 logseq 里用 `[[` 创建一个page的链接的时候，跑一个quicker脚本，通过 Obsidian Advanced URI 新建该page名的笔记，然后在 Obsidian 里撰写或剪藏。然后再回到 Logseq 里，是可以通过双链正常md渲染访问到这个笔记的，而且也能同时维持 obsidian 和 logseq 两个平台的双链关系。

简单地说就是logseq专门负责每日笔记和日志， Obsidian专门负责知识库和文档。

**Reply by 1wingedangel** [reply]
> obsidian是![ ] typora是![img] 能请教叫下这个quicker里应该如何修改吗 ![[^]]*]((http(s)?://.+?.(jpg|jpeg|png|gif))) 我改成这个不行

我发布了新的脚本，改动了很多地方所以干脆重新发布了新的动作脚本了……[地址](https://getquicker.net/Sharedaction?code=16d0410b-2e88-4e30-91fb-08d982821308)

主要区别在于：
* 第一次运行时使用状态变量来保存工作文件夹路径，而不再需要创建以及指定环境变量；
* 工作文件夹路径最好为保存md文档的路径；
* 需要额外配置 Obsidian 。（把 “复制当前文件的路径” 的快捷键改成 Ctrl+Shift+Y ）
* 能够识别的md语法增加了。包括带label的，url后面带参数的，图片预览外面再嵌了个Link的……
* 下载好的图片的位置改为：`工作文件夹\Assets\笔记名称\yyyyMMddHHmmss_ffff.jpg` 方便迁移md文档以及兼容其他md编辑器

大致的原理和用法写在了该动作的描述中。本文内容已经过时，我找个时间清理一下吧……

**Reply by Kenshin** [reply]
嗨，有时间给新版写教程，可以再开个 discussions

@1wingedangel

**Reply by 1wingedangel** [reply]
新脚本教程已更新。原来的脚本与教程作为存档放在了文末。

**Comment by 1wingedangel** [comment]
Obsidian的社区插件市场上已经有了类似的插件：[aleksey-rezvov/obsidian-local-images](https://github.com/aleksey-rezvov/obsidian-local-images)

如果没有什么特殊的需求的话，你们可以用插件试一试。

**Reply by yzg1943** [reply]
这个插件下载公众号的文章，经常重名互相覆盖...

**Comment by SAKURAGI-1983** [comment]
![image](https://user-images.githubusercontent.com/92950903/138472305-e6d3b1f3-41ef-423d-adcb-d6b8b7fd07c5.png)

![image](https://user-images.githubusercontent.com/92950903/138472402-f21affaf-100f-4b91-a8e6-8568ff83cc92.png)


老哥这个是调试运行中出现问题的部分，我都是按步骤来的，哪里出了问题啊。。

**Reply by 1wingedangel** [reply]
obsidian的复制笔记的地址的快捷键没有设置好

**Reply by SAKURAGI-1983** [reply]
> obsidian的复制笔记的地址的快捷键没有设置好

![image](https://user-images.githubusercontent.com/92950903/138491226-e1eff8b9-92ce-4f64-ad78-3a92d73a3c76.png)

老哥 改过快捷键后下边那三条红色信息还是在的，之前截图那张是故意改掉快捷键看看手动输入会不会好使

**Reply by 1wingedangel** [reply]
看截图很明显，是因为剪贴板里的内容不是obsidian笔记的位置，你可能运行脚本时还是没有触发obsidian上的复制文档位置的快捷键。

手动输入的方式可能有点bug，找个时间我去研究一下。

**Reply by SAKURAGI-1983** [reply]
![image](https://user-images.githubusercontent.com/92950903/138553121-84573909-4358-4886-be3b-280279bcbe94.png)
![image](https://user-images.githubusercontent.com/92950903/138553131-7955f320-6571-489a-9312-5a18db215f76.png)

排除了一下全局快捷键占用，更改一下快捷键依然会出现最后三行的标红文字。

具体试验是选中or复制了一段文字+一张图片链接，执行动作会出现
![image](https://user-images.githubusercontent.com/92950903/138553226-7b8d7fb8-dc63-4a5e-b6c8-624f8e9012db.png)
![image](https://user-images.githubusercontent.com/92950903/138553246-eb2752c6-5503-4759-aca6-614a1e2b8a25.png)

此时粘贴板会出现这篇文章的标题，粘贴后如下
![image](https://user-images.githubusercontent.com/92950903/138553338-3b194b2b-8d12-4204-996d-19e99c772aac.png)

**Reply by 1wingedangel** [reply]
感觉像是Obsidian出发的复制文档位置的动作出现了严重的延时。脚本上有做过3秒检查剪贴板是否有变动，看起来在你的环境下输入快捷键→复制文档位置到剪贴板这个步骤超过了三秒。但是不知道你这个md文件是否在根目录下，所以我这个猜测也不太能确定。

你如果方便的话，把调试运行后的htm文件发给我看一下。

P.S. 在运行这个脚本的时候，请确保带有图片预览代码的文本行里没有其他文本内容，不然运行脚本之后这些文本是会被覆盖掉的。在运行脚本前请在文本内容前插入换行符预处理以下再运行。比如：

![image](https://user-images.githubusercontent.com/396190/138577666-8d551134-788f-4c98-bd35-9736b51be88d.png)

**Reply by SAKURAGI-1983** [reply]
脚本的原始文件我只尝试过动一下复制路径的快捷键，因为是开始怀疑是其他求quicker的动作给占用了，文件库是是在D盘新建了一个 markdown 文件夹作为库，我这次是把全文删除只留这两段并进行换行，

![image](https://user-images.githubusercontent.com/92950903/138620471-a9aa91c2-19f4-4685-8b75-cd6b6dafc05a.png)

粘贴后还是只出现文档名称

![image](https://user-images.githubusercontent.com/92950903/138620594-5df3a420-600e-4297-a7cd-26a01e403e3a.png)

以下是html文件

[测试.md](https://github.com/Kenshin/simpread/files/7406622/default.md)

**Reply by 1wingedangel** [reply]
我知道问题在哪里了。

我在正则提取规则里用了零宽断言来保证获取笔记名称时不包含其父目录，导致当笔记位置在 Vault 的根目录下的时候因没有父目录而匹配失败。

我已更新了脚本，你再试试看。

**Reply by SAKURAGI-1983** [reply]
我好使了！

给后边的小伙伴 - 

第一次运行quicker插件时需要指定一个根目录，要指定为你文件库内的 Assets 文件夹，这样链接改变以后ob会跳出来一个跟你md文件同名的文件夹，并且下面还会带有 Assets 文件夹，图片会存储到里面

![image](https://user-images.githubusercontent.com/92950903/138692944-0d1f7f9c-b3a5-45ef-89ad-309619e9fd66.png)

![image](https://user-images.githubusercontent.com/92950903/138692994-fd034d7a-9d10-404c-ae36-5211fd0fee4e.png)

如果第一次运行quiker时指定是库文件夹，比如我的文件夹是 D：markdown ，此时会出现下面的框，并且图片不会下载成功

![image](https://user-images.githubusercontent.com/92950903/138634396-d73eb2f7-c1a7-4c43-b532-e2484a9fabc8.png)

感觉肯定会有和我一样的小白选手看到这儿，如果有错老哥指正一下，老哥方便的话可以加我ID名，一点心意万分感谢~

2021年10月25日 20:23:41

**Reply by 1wingedangel** [reply]
怎么感觉在你的环境下跑出来的脚本还是不太正常呢…

你如果方便的话加一下简悦的discord群，我们讨论一下… id 是 `Vergilius Lee#1460`

**Comment by YB-Kang** [comment]


**Reply by 1wingedangel** [reply]
1. 执行下图的命令，即可清空状态：
![image](https://user-images.githubusercontent.com/396190/139769036-894931e2-754b-4566-8ee6-7e13d43f0b71.png)
2. 自行修改`下载文件`命令上的路径到自己想要的位置上。不难。

**Comment by YB-Kang** [comment]
![image](https://user-images.githubusercontent.com/55050981/141689311-b461480e-7a2d-4b14-ae67-1a31688e71ee.png)
![image](https://user-images.githubusercontent.com/55050981/141689284-9e3edecd-137b-40b0-87f7-2253f3e7e36b.png)

老哥 这个问题怎么解决？
笔记路径是复制成功的 不过没办法分离出笔记名
我看楼上那个例子好像也是这个问题 但是我没看懂怎么解决

**Reply by 1wingedangel** [reply]
提供一下这个图片的URL地址（OCR质量太差）

或许你看看正则表达式有什么问题导致他匹配不上？

**Reply by YB-Kang** [reply]
[050402-823_log.md](https://github.com/Kenshin/simpread/files/7537333/050402-823_log.md)
我把htm调试文件发给你看下。

**Reply by 1wingedangel** [reply]
看了一下，这个图片的URL比较特殊，无法匹配目前的正则表达式规则。

然而如果就算是调整了正则表达式，由于无法通过quicker来自动判断他的扩展名（你别看URL上写的是JPEG格式，实际上下下来扩展名是 `.jfif` ），所以恐怕通过quicker脚本是无法匹配你这篇文章的情况。

**Reply by YB-Kang** [reply]
老哥，上面那个问题解决了，确实发现有的链接转换不了。

你再帮我看看下面这个问题怎么解决，我在其他obsidian库中转换的图片都被下载到 SimpRead 文件夹下面。

![动画](https://user-images.githubusercontent.com/55050981/141937119-c1bcd5c3-f6d1-45d1-b351-035c2af4bc14.gif)

![image](https://user-images.githubusercontent.com/55050981/141938422-1e42d6f6-3957-48d3-a7ff-2731e675c7d8.png)

然后我在这里改了路径还是依然下载在 SimpRead 下面。删除动作数据试过了，重新安装也试过了，没让我重新选择路径。

**Reply by 1wingedangel** [reply]
脚本运行的日志还是原来那个吗？

我看了一下你的原来的运行日志，你的系统环境变量 `OBWEB` 里设置的就是simpread的文件夹……这个脚本会先读是不是有这个环境变量，有的话就直接才用环境变量里的路径了。你可以删除这个系统环境变量再试试看。

**Reply by YB-Kang** [reply]
我把那条环境变量删除了也没办法啊

**Reply by 1wingedangel** [reply]
调试日志发出来看看……

**Reply by YB-Kang** [reply]
[测试日志.md](https://github.com/Kenshin/simpread/files/7546345/default.md)

我发的这个调试日志是转换成功的，就是下载图片的根目录总是自动选择【E:\A10Nutstore\SimpRead】，其他后面的【\Assets\……】这段没问题。
我尝试把系统变量【OBWEB】删除不行。
然后我把它改成我要的【E:\A10Nutstore\ObsidianA】，再重启电脑还是不行。

**Reply by YB-Kang** [reply]
@1wingedangel 老哥，我刚才折腾了一会儿，问题解决了。我这几天反复重装脚本，也改过环境变量，都没能成功更改默认下载路径。我猜测应该是 Quicker 的问题，需要先删脚本，然后不要立刻安装，把环境变量改了或删了，再重启一下 Quicker ，最后才安装脚本。只有这样才能更改下载路径。

**Reply by 1wingedangel** [reply]
修改环境变量后，是一定要重启quicker，quicker脚本里的环境变量才会更新…

![Screenshot_20211117064752.jpg](https://user-images.githubusercontent.com/396190/142078597-4f6894e8-d693-41a4-8e42-4bf55a880e5b.jpg)

**Reply by YB-Kang** [reply]
记住了，谢谢。
